core/src/test/scala/unit/kafka/admin/TopicCommandWithAdminClientTest.scala

```diff
@@ -0,0 +1,564 @@
+import org.junit.{After, Before, Rule, Test}
+import org.junit.rules.TestName
+
+import scala.collection.JavaConverters._
+import scala.concurrent.ExecutionException
+import scala.util.Random
+
+class TopicCommandWithAdminClientTest extends KafkaServerTestHarness with Logging with RackAwareTest {
+
+  /**
+    * Implementations must override this method to return a set of KafkaConfigs. This method will be invoked for every
+    * test and should not reuse previous configurations unless they select their ports randomly when servers are started.
+    */
+  override def generateConfigs: Seq[KafkaConfig] = TestUtils.createBrokerConfigs(
+    numConfigs = 6,
+    zkConnect = zkConnect,
+    rackInfo = Map(0 -> "rack1", 1 -> "rack2", 2 -> "rack2", 3 -> "rack1", 4 -> "rack3", 5 -> "rack3"
+    )).map(KafkaConfig.fromProps)
+
+  private var topicService: AdminClientTopicService = _
+  private var adminClient: JAdminClient = _
+  private var testTopicName: String = _
+
+  private val _testName = new TestName
+  @Rule def testName = _testName
+
+  def assertExitCode(expected: Int, method: () => Unit) {
+    def mockExitProcedure(exitCode: Int, exitMessage: Option[String]): Nothing = {
+      assertEquals(expected, exitCode)
+      throw new RuntimeException
+    }
+    Exit.setExitProcedure(mockExitProcedure)
+    try {
+      intercept[RuntimeException] {
+        method()
+      }
+    } finally {
+      Exit.resetExitProcedure()
+    }
+  }
+
+  def assertCheckArgsExitCode(expected: Int, options: TopicCommandOptions) {
+    assertExitCode(expected, options.checkArgs)
+  }
+
+  @Before
+  def setup() {
+    // create adminClient
+    val props = new Properties()
+    props.put(CommonClientConfigs.BOOTSTRAP_SERVERS_CONFIG, brokerList)
+    adminClient = JAdminClient.create(props)
+    topicService = AdminClientTopicService(adminClient)
+    testTopicName = s"${testName.getMethodName}-${Random.alphanumeric.take(10).mkString}"
+  }
+
+  @After
+  def close(): Unit = {
+    // adminClient is closed by topicService
+    if (topicService != null)
+      topicService.close()
+  }
+
+  @Test
+  def testParseAssignment(): Unit = {
+    val actualAssignment = TopicCommand.parseReplicaAssignment("5:4,3:2,1:0")
+    val expectedAssignment = Map(0 -> List(5, 4), 1 -> List(3, 2), 2 -> List(1, 0))
+    assertEquals(expectedAssignment, actualAssignment)
+  }
+
+  @Test
+  def testParseAssignmentDuplicateEntries(): Unit = {
+    intercept[AdminCommandFailedException] {
+      TopicCommand.parseReplicaAssignment("5:5")
+    }
+  }
+
+  @Test
+  def testParseAssignmentPartitionsOfDifferentSize(): Unit = {
+    intercept[AdminOperationException] {
+      TopicCommand.parseReplicaAssignment("5:4:3,2:1")
+    }
+  }
+
+  @Test
+  def testInvalidConfigOptWithBootstrapServers(): Unit = {
+    assertCheckArgsExitCode(1,
+      new TopicCommandOptions(Array("--bootstrap-server", brokerList ,"--alter", "--topic", testTopicName, "--config", "cleanup.policy=compact")))
```

Comment:
This will fail with below error. But error should be " --config"  is not allowed.
``` 
>>./kafka-topics.sh --bootstrap-server localhost:9092  --alter --topic foo --config cleanup.policy=compact
Missing required argument "[partitions]"

//Below zk based command should work

>>./kafka-topics.sh --zookeeper localhost:2181 --alter --topic foo --config cleanup.policy=compact
Missing required argument "[partitions]"
```

Reply:
Well, in this case I think I'm gonna fix the test as the partitions opt should be required for the `alter` option in this case since `alter` is reduced to the functionality of changing the partition count.